File: src/main/java/net/guizhanss/ultimategenerators2/implementation/items/abstracts/AMachine.java
package net.guizhanss.ultimategenerators2.implementation.items.abstracts;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.annotation.ParametersAreNonnullByDefault;

import com.google.common.base.Preconditions;

import org.bukkit.Location;
import org.bukkit.block.Block;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.inventory.ItemStack;

import io.github.thebusybiscuit.slimefun4.api.SlimefunAddon;
import io.github.thebusybiscuit.slimefun4.api.items.ItemGroup;
import io.github.thebusybiscuit.slimefun4.api.items.ItemState;
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItemStack;
import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType;
import io.github.thebusybiscuit.slimefun4.core.attributes.EnergyNetComponent;
import io.github.thebusybiscuit.slimefun4.core.attributes.MachineProcessHolder;
import io.github.thebusybiscuit.slimefun4.core.machines.MachineProcessor;
import io.github.thebusybiscuit.slimefun4.core.networks.energy.EnergyNetComponentType;
import io.github.thebusybiscuit.slimefun4.implementation.operations.CraftingOperation;
import io.github.thebusybiscuit.slimefun4.libraries.dough.inventory.InvUtils;
import io.github.thebusybiscuit.slimefun4.utils.SlimefunUtils;
import io.github.thebusybiscuit.slimefun4.utils.itemstack.ItemStackWrapper;

import me.mrCookieSlime.Slimefun.api.BlockStorage;
import me.mrCookieSlime.Slimefun.api.inventory.BlockMenu;

import net.guizhanss.ultimategenerators2.api.recipes.MachineRecipe;
import net.guizhanss.ultimategenerators2.api.recipes.SingleMachineRecipe;
import net.guizhanss.ultimategenerators2.api.recipes.SingleOutputMachineRecipe;

import lombok.Getter;

public abstract class AMachine extends AHopper implements EnergyNetComponent, MachineProcessHolder<CraftingOperation> {
    protected final List<MachineRecipe> recipes = new ArrayList<>();
    private final MachineProcessor<CraftingOperation> processor = new MachineProcessor<>(this);

    @Getter
    private int energyCapacity = -1;
    @Getter
    private int energyConsumption = -1;

    @ParametersAreNonnullByDefault
    protected AMachine(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe) {
        super(itemGroup, item, recipeType, recipe);

        processor.setProgressBar(getProgressBar());
    }

    public abstract ItemStack getProgressBar();

    @Override
    @Nonnull
    public final MachineProcessor<CraftingOperation> getMachineProcessor() {
        return processor;
    }

    @Override
    @ParametersAreNonnullByDefault
    protected void onBreak(BlockBreakEvent e, BlockMenu menu) {
        super.onBreak(e, menu);
        processor.endOperation(menu.getLocation());
    }

    @Nonnull
    @Override
    public final EnergyNetComponentType getEnergyComponentType() {
        return EnergyNetComponentType.CONSUMER;
    }

    @Override
    public int getCapacity() {
        return energyCapacity;
    }

    @Nonnull
    public final AMachine setCapacity(int capacity) {
        Preconditions.checkArgument(capacity > 0, "Capacity must be greater than 0");

        if (getState() == ItemState.UNREGISTERED) {
            this.energyCapacity = capacity;
            return this;
        } else {
            throw new IllegalStateException("You cannot modify the capacity after the Item was registered.");
        }
    }

    @Nonnull
    public final AMachine setEnergyConsumption(int energyConsumption) {
        Preconditions.checkArgument(energyConsumption > 0, "Energy consumption must be greater than 0");

        if (getState() == ItemState.UNREGISTERED) {
            this.energyConsumption = energyConsumption;
            return this;
        } else {
            throw new IllegalStateException("You cannot modify the energy consumption after the Item was registered.");
        }
    }

    protected abstract void registerDefaultRecipes();

    @Nonnull
    @Override
    public List<ItemStack> getDisplayRecipes() {
        List<ItemStack> list = new ArrayList<>();
        for (MachineRecipe recipe : recipes) {
            int length = Math.max(recipe.getInput().length, recipe.getOutput().length);
            ItemStack[] inputs = Arrays.copyOf(recipe.getInput(), length);
            ItemStack[] outputs = Arrays.copyOf(recipe.getOutput(), length);
            for (int i = 0; i < length; i++) {
                list.add(inputs[i]);
                list.add(outputs[i]);
            }
        }
        return list;
    }

    @ParametersAreNonnullByDefault
    protected void registerRecipe(int ticks, ItemStack[] inputs, ItemStack[] outputs) {
        recipes.add(new MachineRecipe(ticks, inputs, outputs));
    }

    @ParametersAreNonnullByDefault
    protected void registerRecipe(int ticks, ItemStack[] inputs, ItemStack outputs) {
        recipes.add(new SingleOutputMachineRecipe(ticks, inputs, outputs));
    }

    @ParametersAreNonnullByDefault
    protected void registerRecipe(int ticks, ItemStack input, ItemStack output) {
        recipes.add(new SingleMachineRecipe(ticks, input, output));
    }

    @Override
    @ParametersAreNonnullByDefault
    protected void tick(Block b, BlockMenu menu) {
        BlockMenu inv = BlockStorage.getInventory(b);
        CraftingOperation currentOperation = processor.getOperation(b);

        if (currentOperation != null) {
            if (takeCharge(b.getLocation())) {
                if (!currentOperation.isFinished()) {
                    processor.updateProgressBar(inv, STATUS_SLOT, currentOperation);
                    currentOperation.addProgress(1);
                } else {
                    inv.replaceExistingItem(STATUS_SLOT, NOT_WORKING_ITEM);

                    for (ItemStack output : currentOperation.getResults()) {
                        inv.pushItem(output.clone(), getOutputSlots());
                    }

                    processor.endOperation(b);
                }
            }
        } else {
            MachineRecipe next = findNextRecipe(inv);

            if (next != null) {
                processor.startOperation(b, new CraftingOperation(next.toSf()));
            }
        }
    }

    protected boolean takeCharge(@Nonnull Location l) {
        Preconditions.checkNotNull(l, "Location cannot be null");

        if (isChargeable()) {
            int charge = getCharge(l);

            if (charge < getEnergyConsumption()) {
                return false;
            }

            setCharge(l, charge - getEnergyConsumption());
        }
        return true;
    }

    @Nullable
    protected MachineRecipe findNextRecipe(@Nonnull BlockMenu inv) {
        Map<Integer, ItemStack> inventory = new HashMap<>();

        for (int slot : getInputSlots()) {
            ItemStack item = inv.getItemInSlot(slot);

            if (item != null) {
                inventory.put(slot, ItemStackWrapper.wrap(item));
            }
        }

        Map<Integer, Integer> found = new HashMap<>();

        for (MachineRecipe recipe : recipes) {
            for (ItemStack input : recipe.getInput()) {
                for (int slot : getInputSlots()) {
                    if (SlimefunUtils.isItemSimilar(inventory.get(slot), input, false)) {
                        found.put(slot, input.getAmount());
                        break;
                    }
                }
            }

            if (found.size() == recipe.getInput().length) {
                if (!InvUtils.fitAll(inv.toInventory(), recipe.getOutput(), getOutputSlots())) {
                    return null;
                }

                for (Map.Entry<Integer, Integer> entry : found.entrySet()) {
                    inv.consumeItem(entry.getKey(), entry.getValue());
                }

                return recipe;
            } else {
                found.clear();
            }
        }

        return null;
    }

    @Override
    public void register(@Nonnull SlimefunAddon addon) {
        Preconditions.checkArgument(energyCapacity > 0, "Energy capacity must be greater than 0");
        Preconditions.checkArgument(energyConsumption > 0, "Energy consumption must be greater than 0");
        Preconditions.checkArgument(energyConsumption <= energyCapacity, "Energy consumption must be less than or equal to the energy capacity");
        super.register(addon);
        registerDefaultRecipes();
    }
}

File: src/main/kotlin/net/guizhanss/ultimategenerators2/implementation/items/abstracts/AMachine.kt
package net.guizhanss.ultimategenerators2.implementation.items.abstracts

import com.google.common.base.Preconditions
import io.github.thebusybiscuit.slimefun4.api.SlimefunAddon
import io.github.thebusybiscuit.slimefun4.api.items.ItemGroup
import io.github.thebusybiscuit.slimefun4.api.items.ItemState
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItemStack
import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType
import io.github.thebusybiscuit.slimefun4.core.attributes.EnergyNetComponent
import io.github.thebusybiscuit.slimefun4.core.attributes.MachineProcessHolder
import io.github.thebusybiscuit.slimefun4.core.machines.MachineProcessor
import io.github.thebusybiscuit.slimefun4.core.networks.energy.EnergyNetComponentType
import io.github.thebusybiscuit.slimefun4.implementation.operations.CraftingOperation
import io.github.thebusybiscuit.slimefun4.libraries.dough.inventory.InvUtils
import io.github.thebusybiscuit.slimefun4.utils.SlimefunUtils
import io.github.thebusybiscuit.slimefun4.utils.itemstack.ItemStackWrapper
import me.mrCookieSlime.Slimefun.api.BlockStorage
import me.mrCookieSlime.Slimefun.api.inventory.BlockMenu
import net.guizhanss.ultimategenerators2.core.recipes.MachineRecipe
import net.guizhanss.ultimategenerators2.core.recipes.SingleMachineRecipe
import net.guizhanss.ultimategenerators2.core.recipes.SingleOutputMachineRecipe
import net.guizhanss.ultimategenerators2.utils.BlockUtils
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.inventory.ItemStack
import kotlin.math.max

/**
 * The basic class of a machine.
 */
abstract class AMachine(
    itemGroup: ItemGroup,
    item: SlimefunItemStack,
    recipeType: RecipeType,
    recipe: Array<ItemStack?>
) : AHopper(itemGroup, item, recipeType, recipe), EnergyNetComponent, MachineProcessHolder<CraftingOperation> {
    @JvmField
    protected val recipes: MutableList<MachineRecipe> = ArrayList()

    @JvmField
    protected val processor: MachineProcessor<CraftingOperation> = MachineProcessor(this)

    private var _energyCapacity = -1
    private var _energyConsumption = -1

    @get:JvmName("getEnergyCapacity")
    val capacity: Int get() = _energyCapacity
    val energyConsumption: Int get() = _energyConsumption

    protected abstract val progressBar: ItemStack

    init {
        val lazyProgressBar by lazy { progressBar }
        processor.progressBar = lazyProgressBar
    }

    override fun getCapacity() = capacity

    override fun getMachineProcessor() = processor

    override fun onBreak(e: BlockBreakEvent, menu: BlockMenu) {
        super.onBreak(e, menu)
        processor.endOperation(menu.location)
    }

    override fun getEnergyComponentType() = EnergyNetComponentType.CONSUMER

    fun setCapacity(capacity: Int): AMachine {
        Preconditions.checkArgument(capacity > 0, "Capacity must be greater than 0")

        if (state == ItemState.UNREGISTERED) {
            this._energyCapacity = capacity
            return this
        } else {
            error("You cannot modify the capacity after the Item was registered.")
        }
    }

    fun setEnergyConsumption(energyConsumption: Int): AMachine {
        Preconditions.checkArgument(energyConsumption > 0, "Energy consumption must be greater than 0")

        if (state == ItemState.UNREGISTERED) {
            this._energyConsumption = energyConsumption
            return this
        } else {
            error("You cannot modify the energy consumption after the Item was registered.")
        }
    }

    protected abstract fun registerDefaultRecipes()

    override fun getDisplayRecipes(): List<ItemStack> {
        val list: MutableList<ItemStack> = ArrayList()
        for (recipe in recipes) {
            val length = max(recipe.input.size, recipe.output.size)
            val inputs: Array<ItemStack?> = recipe.input.copyOf(length)
            val outputs: Array<ItemStack?> = recipe.output.copyOf(length)
            for (i in 0 until length) {
                list.add(inputs[i] ?: ItemStack(Material.AIR))
                list.add(outputs[i] ?: ItemStack(Material.AIR))
            }
        }
        return list
    }

    protected fun registerRecipe(ticks: Int, inputs: Array<ItemStack>, outputs: Array<ItemStack>) {
        recipes.add(MachineRecipe(ticks, inputs, outputs))
    }

    protected fun registerRecipe(ticks: Int, inputs: Array<ItemStack>, output: ItemStack) {
        recipes.add(SingleOutputMachineRecipe(ticks, inputs, output))
    }

    protected fun registerRecipe(ticks: Int, input: ItemStack, output: ItemStack) {
        recipes.add(SingleMachineRecipe(ticks, input, output))
    }

    override fun tick(b: Block, menu: BlockMenu) {
        val inv: BlockMenu = BlockStorage.getInventory(b)
        val currentOperation: CraftingOperation? = processor.getOperation(b)

        if (currentOperation != null) {
            if (BlockUtils.takeCharge(this, b.location, energyConsumption)) {
                if (!currentOperation.isFinished) {
                    processor.updateProgressBar(inv, STATUS_SLOT, currentOperation)
                    currentOperation.addProgress(1)
                } else {
                    inv.replaceExistingItem(STATUS_SLOT, NOT_WORKING_ITEM)

                    for (output in currentOperation.results) {
                        inv.pushItem(output.clone(), *outputSlots)
                    }

                    processor.endOperation(b)
                }
            }
        } else {
            val next = findNextRecipe(inv)

            if (next != null) {
                processor.startOperation(b, CraftingOperation(next.toSf()))
            }
        }
    }

    protected fun findNextRecipe(inv: BlockMenu): MachineRecipe? {
        val inventory: MutableMap<Int, ItemStack> = HashMap()

        for (slot in inputSlots) {
            val item: ItemStack? = inv.getItemInSlot(slot)

            if (item != null) {
                inventory[slot] = ItemStackWrapper.wrap(item)
            }
        }

        val found: MutableMap<Int, Int> = HashMap()

        for (recipe in recipes) {
            for (input in recipe.input) {
                for (slot in inputSlots) {
                    if (SlimefunUtils.isItemSimilar(inventory[slot], input, false)) {
                        found[slot] = input.amount
                        break
                    }
                }
            }

            if (found.size == recipe.input.size) {
                if (!InvUtils.fitAll(inv.toInventory(), recipe.output, *outputSlots)) {
                    return null
                }

                for ((key, value) in found) {
                    inv.consumeItem(key, value)
                }

                return recipe
            } else {
                found.clear()
            }
        }

        return null
    }

    override fun register(addon: SlimefunAddon) {
        Preconditions.checkArgument(_energyCapacity > 0, "Energy capacity must be greater than 0")
        Preconditions.checkArgument(_energyConsumption > 0, "Energy consumption must be greater than 0")
        Preconditions.checkArgument(
            _energyConsumption <= _energyCapacity,
            "Energy consumption must be less than or equal to the energy capacity"
        )
        super.register(addon)
        registerDefaultRecipes()
    }
}
